src/coroutine.rs

```diff
@@ -0,0 +1,314 @@
+            FUTURE_UNINITIALIZED,
+            FUTURE_INITIALIZED,
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+        );
+        Ok(if mark_future_as_initialized.is_ok() {
+            // return future if its initialization has be confirmed
+            // SAFETY: state is `FUTURE_INITIALIZED`
+            Some(unsafe { (*self.future.get()).assume_init_ref() }.as_ref(py))
+        } else {
+            // otherwise, waker should have been waken before,
+            // so drop the previously written future and return None
+            debug_assert_eq!(
+                mark_future_as_initialized,
+                Err(WAKEN_BEFORE_FUTURE_INITIALIZED)
+            );
+            // SAFETY: state is `FUTURE_INITIALIZED`
+            unsafe { (*self.future.get()).assume_init_drop() };
+            None
+        })
+    }
+}
+
+impl Drop for AsyncioWaker {
+    fn drop(&mut self) {
+        self.reset()
+    }
+}
+
+// SAFETY: future initialization is synchronized with state
+unsafe impl Send for AsyncioWaker {}
+// SAFETY: future initialization is synchronized with state
+unsafe impl Sync for AsyncioWaker {}
+
+impl futures_task::ArcWake for AsyncioWaker {
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        // if future is uninitialized yet, mark the waker as waken and return
+        let mark_future_as_waken = arc_self.state.compare_exchange(
+            FUTURE_UNINITIALIZED,
+            WAKEN_BEFORE_FUTURE_INITIALIZED,
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+        );
+        match mark_future_as_waken {
+            Ok(_) => return,
+            Err(s) => debug_assert_eq!(s, FUTURE_INITIALIZED),
+        }
+        // otherwise, choose appropriate wake method depending on whether it is called
+        // on the same thread than the event loop
+        // SAFETY: waker state is `FUTURE_INITIALIZED`
+        let future = unsafe { (*arc_self.future.get()).assume_init_ref() };
+        Python::with_gil(|gil| future_set_result_threadsafe(future.as_ref(gil)))
+    }
+}
+
+/// Instantiate an `asyncio.Future` in the current event loop.
+fn asyncio_future(py: Python<'_>) -> PyResult<PyObject> {
+    static FUTURE: GILOnceCell<PyObject> = GILOnceCell::new();
+    let import_future = || -> PyResult<_> {
+        let module = py.import("asyncio")?;
+        Ok(module.getattr("Future")?.into())
+    };
+    FUTURE.get_or_try_init(py, import_future)?.call0(py)
+}
+
+/// Set the result of an `asyncio.Future` (thread safe)
+fn future_set_result_threadsafe(future: &PyAny) {
+    // `set_result` must be called in event loop thread, so it requires `call_soon_threadsafe`
+    let set_result = future
+        .getattr(intern!(future.py(), "set_result"))
+        .expect("error while calling Future.set_result");
+    let event_loop = future
+        .call_method0(intern!(future.py(), "get_loop"))
+        .expect("error while calling Future.get_loop");
+    let call_soon_threadsafe = event_loop.call_method1(
```

Comment:
A future is normally bound to one event loop (and its `_loop` attribute is private), so I think its safe to cache the method.

Reply:
Actually, the coroutine may yield several futures, and if you want to make awful things, you could run the coroutine manually (calling `send`) in one event loop, and then in another.

However, I'm not sure that caching the method is a good idea in the end, because I think that most of the Rust futures wrapped should only be waken once; if you have a bigger future with several of suspension points, you may want to spawn a task in a Rust runtime to have the future polled on Rust side.
I don't really want to penalize small futures with an additional allocation for the cached method handle, so I prefer to keep the method call.

By the way, calling directly `asyncio.Future()` was a mistake, because `EventLoop.create_future()` is the recommended method. I will fix this.

Reply:
Fair enough, the resulting simplicity is also a nice benefit of not trying to memoize anything.